### PR TITLE
Creating nuclear data using NJOY

### PR DESCRIPTION
**Pull request by paulromano**
This pull request introduces a new capbility to generate an `IncidentNeutron` or `ThermalScattering` object using NJOY (which is now [open source](https://github.com/njoy/NJOY2016)) by adding `from_njoy` classmethods. These functionally work as though they were `from_endf` because they accept the name of an ENDF file, but under the hood, NJOY is run to produce ACE files (at multiple temperatures if desired) and then `from_ace` is used to instantiate the corresponding incident neutron or thermal scattering data. The new methods are also "smart" in that they will automatically add fission energy release data from the ENDF file as well as 0 K elastic scattering data (using the pointwise ENDF file generated during the NJOY run).

Implementation: there is a new `openmc.data.njoy` interface which provides `make_pendf`, `make_ace` and `make_ace_thermal` functions. All of these use a central `run` function. The `run` function has optional arguments `njoy_exec` (to specify the location of the NJOY executable -- assumed to just be 'njoy' by default) and `stdout` (if you want to see the progress of the NJOY run). Keyword arguments are passed through the series of calls from `IncidentNeutron.from_njoy` to `make_ace` to `run`, so, e.g., you can call
```
u235 = openmc.data.IncidentNeutron.from_njoy('U235.endf', stdout=True)
```
Right now, reconstruction/broadening error tolerances are not configurable, but we can change that if need be.

With this functionality, it is now *much* easier for a user to generate an HDF5 library for use in OpenMC given a set of ENDF files.

### CODE REVIEW

File: openmc/data/thermal.py [hunk 1]
```diff
@@ -582,3 +586,47 @@ def from_ace(cls, ace_or_filename, name=None):
...
+            Keyword arguments passed to :func:`openmc.data.njoy.make_ace_thermal`
+
+        Returns
+        -------
+        data : openmc.data.ThermalScattering
+            Thermal scattering data
+
+        """
+        # Create temporary directory -- it would be preferable to use
+        # TemporaryDirectory(), but it is only available in Python 3.2
```
Comment (wbinventor) [inline openmc/data/thermal.py]: Do you plan to migrate to using `TemporaryDirectory` some day when Python 2 dies? If not, this comment may not be so useful to keep around. Just wasn't sure if this was simply a note to yourself during development, or a note to yourself for the future.

**Comment by wbinventor** [comment]
This PR looks good to me, though I'm not particularly qualified nor have I tried any of this code out. Perhaps @smharper or @nelsonag may want to take a look as well. Also, do you think it worthwhile to add a new notebook showcasing these new features?

File: openmc/data/thermal.py (hunk 1 above)
Comment (paulromano) [inline openmc/data/thermal.py]: Yes, by the time Python 2 dies (2020) or before, I would advocate switching over to a Py3-only codebase.

**Comment by paulromano** [comment]
Thanks for the suggestion -- I'll update the nuclear data notebook to mention this capability.

**Comment by paulromano** [comment]
Also added a section in the user's manual on the NJOY capabilities.

**Comment by nelsonag** [comment]
I haven't really used NJOY much so sadly I can't weigh in anymore than
@wbinventor did. I guess I have a bigger picture question though: did you
(@paulromano) try and use this to regenerate the NNDC ENDF/B-VII.0 data?
Does it match, if not are differences understood?

**Comment by nelsonag** [comment]
Nm that would be futile. How about just comparing the input decks produced
with this

**Comment by paulromano** [comment]
I could do a full comparison but it would be rather laborious and probably not that insightful. I don't have access to the input decks that NNDC used to generate ACE files, but I can compare to the JEFF 3.2 input files, which are available. Taking a look at them now, the only differences I see are:
- For JEFF, they ran `GASPR` to add gas production reactions to the files. We don't use those, so not an issue.
- For JEFF, they also ran `THERMR` (even for non-moderators), but if I understand correctly that is only needed if you are also producing multi-group cross sections via `GROUPR`/`MATXSR`.
- JEFF used a value of 2 MeV for `thnmax` (upper limit of Doppler broadening) when running `BROADR`. We use 1 MeV.
- For JEFF, to generate ACE files at multiple temperatures, they do completely separate runs of NJOY for each temperature. The `from_njoy` methods here take advantage of the fact that most NJOY modules can process multiple temperatures simultaneously (the one exception is ACER, so that has to be run multiple times, but we do it in the same "deck").

**Comment by nelsonag** [comment]
I'm happy with that, thanks @paulromano 
I'll give until tomorrow afternoon for @smharper to chime in if he would like to do a more thorough review.  If nothing by then I will merge.

**Comment by nelsonag** [comment]
Ok, ready to merge! Thanks @paulromano and @wbinventor for the more thorough review!

**Pull request merged**